### PR TITLE
Feature/bt18 add contract spec

### PR DESCRIPTION
**Pull request by hoani**
Add smart contract specification to doughnut paper

* Updated overview payload example
* Add contracts to permissions
* Add contracts to decoding
* Updated permissions to correctly identify constraints

### CODE REVIEW

File: CENNZnet_format.md [hunk 1]
```diff
@@ -59,15 +59,52 @@ The domain is binary encoded. It contains a version definition, and a permission
                 * LE unsigned integer
         * If `constraints` is set
             * 1 byte
-                * Constraints length
-
+                * `constraints_length`
```
Comment (jordy25519) [inline CENNZnet_format.md]: 👍

File: CENNZnet_format.md [hunk 2]
```diff
@@ -59,15 +59,52 @@ The domain is binary encoded. It contains a version definition, and a permission
...
+                * LE unsigned integer
+                * `constraints_length` = byte_value + 1
+            * `constraints_length` bytes
+                * constraints payload
+* Contracts list
+    * 1 byte
+        * 1 bit
+            * `block_cooldown` presence
+        * 7 bits
```
Comment (jordy25519) [inline CENNZnet_format.md:74]: Would prefer to leave methods out for now as methods on contracts may require a different treatment

**Review by jordy25519 — COMMENTED** [review]
Thanks for the tidy ups.
Let's leave contract methods out for now. A whitelist of contract addresses is sufficient

File: CENNZnet_format.md (hunk 2 above)
Comment (hoani) [reply CENNZnet_format.md:74]: Done

File: CENNZnet_format.md [hunk 3]
```diff
@@ -145,7 +122,7 @@ The domain is binary encoded. It contains a version definition, and a permission
     3.  Named module constraints take priority over wildcard module constraints
 2.  Any contract named `"*"` is a wildcard
     1.  A wildcard contract enables all contracts
-    2.  Constraints applied to a wildcard contract such as `methods` and `block_cooldown` apply
+    2.  Constraints applied to a wildcard contract such as `block_cooldown` apply
```
Comment (jordy25519) [inline CENNZnet_format.md:125]: just realized the term 'constraints' is confusing. Should probably change that to say pact bytecode at some point

**Review by jordy25519 — APPROVED** [review]
LGTM 🏅

**Pull request merged**